### ISSUE TITLE
Bug: Array literals that contain function or implicit objects with function values

### ISSUE THREAD

**Issue by naeleejang** [jashkenas/coffeescript#5121]
Choose one: is this a bug report or feature request?

Bug report

### Input Code

```coffee
[k: ->]
```
and
```coffee
[j: -> m is 0]
```
and
```coffee
[->]
```

### Expected Behavior
The above should compile.

### Current Behavior
It does not compile.

### Context
```coffee
[->a]
```
compiles and
```coffee
[->,a]
```
compiles

### Environment

* CoffeeScript version: 2 only, version 1 compiles
* Node.js version:

**Comment by naeleejang** [comment]
```
[->]
```
also does not compile

**Comment by helixbass** [comment]
Ya I think at least the cases where `->` is immediately followed by `]` could be handled by a similar fix to #4588 

With the other example `[j: -> m is 0]`, I'm not sure, you'd probably have to go a little deeper into the rewriter

**Comment by GeoffreyBooth** [comment]
All of these are fixed in 2.5.0 except `[j: -> m is 0]`, which I’m not sure is something that should necessarily be supported; it can be compiled if you use multiline style:

```coffee
[j: ->
  m is 0
]
```
